src/protocol.h

```diff
@@ -368,23 +380,15 @@ class CAddress : public CService
     CAddress() : CService{} {};
     explicit CAddress(CService ipIn, ServiceFlags nServicesIn) : CService{ipIn}, nServices{nServicesIn} {};
 
-    SERIALIZE_METHODS(CAddress, obj)
+    SERIALIZE_METHODS_PARAMS(CAddress, obj, AddrSerialization, fmt)
     {
         SER_READ(obj, obj.nTime = TIME_INIT);
-        if (s.GetType() & SER_DISK) {
+        if (fmt == AddrSerialization::DISK) {
             uint32_t disk_version = DISK_VERSION;
             READWRITE(disk_version);
             READWRITE(obj.nTime);
-        } else if (s.GetType() & SER_NETWORK) {
-            // The only time we serialize a CAddress object without nTime is in
-            // the initial VERSION messages which contain two CAddress records.
-            // At that point, the serialization version is INIT_PROTO_VERSION.
-            // After the version handshake, serialization version is >=
-            // MIN_PEER_PROTO_VERSION and all ADDR messages are serialized with
-            // nTime.
-            if (s.GetVersion() >= INIT_PROTO_VERSION) {
-                READWRITE(obj.nTime);
-            }
+        } else if (fmt == AddrSerialization::NETWORK_WITHTIME) {
+            READWRITE(obj.nTime);
         }
```

Comment:
in commit "user serialization ...":


Still not compatible. Now I get:

```
2020-07-13T17:01:39Z [opencon] trying connection 95.183.48.89:18333 lastseen=2468.3hrs
2020-07-13T17:01:39Z [opencon] Added connection peer=2
2020-07-13T17:01:39Z [opencon] sending version (103 bytes) peer=2
2020-07-13T17:01:39Z [opencon] send version message: version 70015, blocks=1746969, us=[::]:0, peer=2
2020-07-13T17:01:39Z [msghand] received: version (102 bytes) peer=2
2020-07-13T17:01:39Z [msghand] ProcessMessages(version, 102 bytes): Exception 'CDataStream::read(): end of data: iostream error' (NSt8ios_base7failureB5cxx11E) caught
2020-07-13T17:01:39Z [msghand] received: verack (0 bytes) peer=2
2020-07-13T17:01:39Z [msghand] Misbehaving: 95.183.48.89:18333 peer=2 (0 -> 1)

Reply:
There was an unrelated bug to the previous one in this commit, with the same effect: `addrFrom` in the VERSION message was parsed with `NETWORK_WITHTIME`.